simple/rdm_triggeredOps.c

```diff
@@ -0,0 +1,670 @@
+	compare_atomic.compare = compare;
+	compare_atomic.flags = 0;
+	work.op_type = FI_OP_COMPARE_ATOMIC;
+
+	work.op.compare_atomic = &compare_atomic;
+	return fi_control(&domain->fid, FI_QUEUE_WORK, &work);
+}
+
+static int tsend_trigger()
+{
+	int ret;
+	struct fi_op_tagged tagged;
+	struct fi_msg_tagged msg;
+	struct iovec msg_iov;
+
+	format_simple_msg_tagged(&msg, &msg_iov, rx_buf,
+				 strlen(write_text) + 1, &work.context, 0xCAFE);
+	tagged.ep = trig_ep;
+	tagged.msg = msg;
+	tagged.flags = 0;
+	work.op.tagged = &tagged;
+
+	work.op_type = FI_OP_TRECV;
+	ret = fi_control(&domain->fid, FI_QUEUE_WORK, &work);
+	if (ret)
+		return ret;
+
+	work.context = trig_ctx2;
+
+	format_simple_msg_tagged(&msg, &msg_iov, tx_buf + strlen(welcome_text),
+				 strlen(write_text) + 1, &work.context, 0xCAFE);
+	work.op_type = FI_OP_TSEND;
+	return fi_control(&domain->fid, FI_QUEUE_WORK, &work);
+}
+
+static int send_trigger()
+{
+	int ret;
+	struct fi_op_msg op_msg;
+	struct fi_msg msg;
+	struct iovec msg_iov;
+
+	format_simple_msg(&msg, &msg_iov, rx_buf,
+			  strlen(write_text) + 1, &work.context);
+
+	op_msg.ep = trig_ep;
+	op_msg.msg = msg;
+	op_msg.flags = 0;
+	work.op.msg = &op_msg;
+
+	work.op_type = FI_OP_RECV;
+	ret = fi_control(&domain->fid, FI_QUEUE_WORK, &work);
+	if (ret)
+		return ret;
+
+	work.context = trig_ctx2;
+
+	format_simple_msg(&msg, &msg_iov, tx_buf + strlen(welcome_text),
+			  strlen(write_text) + 1, &work.context);
+	work.op_type = FI_OP_SEND;
+	return fi_control(&domain->fid, FI_QUEUE_WORK, &work);
+}
+
+static int trigger()
+{
+	int ret;
+
+	if (opts.dst_addr) {
+		work.triggering_cntr = txcntr;
+		rx_exp = 3;
+	} else {
+		work.triggering_cntr = rxcntr;
+		rx_exp = 4;
+	}
+	work.threshold = 3;
+
+	if (tested_op != FI_OP_CNTR_ADD && tested_op != FI_OP_CNTR_SET)
+		work.completion_cntr = test_cntr;
+
+	switch (tested_op) {
+	case FI_OP_RECV:
+	case FI_OP_SEND:
+		ret = send_trigger();
+		n_trig++;
+		break;
+	case FI_OP_TRECV:
+	case FI_OP_TSEND:
+		ret = tsend_trigger();
+		n_trig++;
+		break;
+	case FI_OP_READ:
+	case FI_OP_WRITE:
+		ret = rma_trigger();
+		n_trig++;
+		rx_exp += 2;
+		break;
+	case FI_OP_ATOMIC:
+		ret = atomic_trigger();
+		rx_exp++;
+		break;
+	case FI_OP_FETCH_ATOMIC:
+		ret = fetch_atomic_trigger();
+		rx_exp++;
+		break;
+	case FI_OP_COMPARE_ATOMIC:
+		ret = compare_atomic_trigger();
+		rx_exp++;
+		break;
+	default:
+		work.op_type = tested_op;
+		ret = cntr_trigger();
+		n_trig++;
+	}
+	n_trig++;
+
+	if (ret) {
+		if (ret == -FI_ENOSYS)
+			fprintf(stdout, "Operation not supported by provider\n");
+		else
+			FT_PRINTERR("fi_control", ret);
+		return ret;
+	}
+
+	ret = send_wait_check();
+
+	return ret;
+}
+
+static void init_buf_vals()
+{
+	switch (tested_op) {
+	case FI_OP_ATOMIC:
+	case FI_OP_FETCH_ATOMIC:
+	case FI_OP_COMPARE_ATOMIC:
+		memset(compare_buf, ~0, strlen(compare_buf));
+		sprintf(tx_buf, "%s", welcome_text);
+		memset(&tx_buf[strlen(welcome_text)], ~0, strlen(welcome_text));
```

Comment:
Note that strlen() does not include the null terminator.  I noticed there are several places where strlen() is used, and others where it's strlen() + 1.  I didn't delve into the details to see if all of the use cases were correct.  :)